File: szakdoliV2/Drawable/Tab.py
# -*- coding: utf-8 -*-

from Polygon import Polygon
from System.Constants import COLOR as Color


#TODO: Tényleg újragondolni az egészet, és az AbstractDrawable osztályból származtatni
class Tab(Polygon):

    def __init__(self, x, y, w, h, color = Color.BLACK , id = None, width = 0, descriptor = "", size = None, vec2_pos = None, transparent=True):
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.id = id
        self.color = color
        self.darkAccent = Color.GREY
        self.lightAccent = (200, 200, 200)    #implicit szinmegadás, már már a magic number határát karistolja
        self.accentColor = None
        self.width = width
        self.descriptor = descriptor
        self.transparent = transparent
        self.selected = False
        self.coordinates = self.CalculatePoints()

    def IsInside(self, position):
        return self.x <= position[0] and self.x + self.w >= position[0] and self.y <= position[1] and self.y + self.h >= position[1]

    def DrawObject(self, screen):
        #pygame.draw.rect(screen, Color.RED, (self.x, self.y, self.w, self.h))
        if not self.transparent:
            if self.selected:
                self.accentColor = self.darkAccent
            else:
                self.accentColor = self.lightAccent
        super(Tab, self).DrawObject(screen)

    def GetId(self):
        return self.id

    def CalculatePoints(self):
        points = []
        percentofcut = 0.2
        '''
        óra mutatóval megegyezően helyezzük fel a pontokat
        a magasság 20%-a lesz a legvágott rész

            *(p2)--------*(p3)      /\              *(tl) ------------ * (tr)       tl = self.x, self.y
          /                \         | percentofcut |                  |            tr = self.x + self.w, self.y
        *(p1)               *(p4)   \/              |                  |            bl = self.x, self.y + self.h
        |                   |                       |                  |            br = self.x + self.w, self.y + self.h
        |                   |                       |                  |
        |                   |                       |                  |
        *(p6) --------------*(p5)                   *(bl) ------------ * (br)
        '''

        # TODO lehetne 4 részből összerakni, 2 téglalap és 2 ARC

        tl = self.x, self.y
        tr = self.x + self.w, self.y
        bl = self.x, self.y + self.h
        br = self.x + self.w, self.y + self.h
        edgecut = int(self.h * percentofcut)

        #fontos hogy int legyen mert csak egész pontokat lehet megjeleníteni
        #      Coords:  X      Y
        points.append((tl[0] , tl[1] + edgecut )) #p1
        points.append((tl[0] + edgecut, tl[1]))  # p2
        points.append((tr[0] - edgecut, tr[1]  ))  # p3
        points.append((tr[0] , tr[1] + edgecut ))  # p4
        points.append( br )  # p5
        points.append( bl )  # p6

        return points

File: docs/Drawable/Tab.js
import { Polygon } from "./Polygon.js";
import { COLOR } from "../System/Constants.js";

export class Tab extends Polygon {

    constructor(x, y, w, h, color = COLOR.BLACK, id = null, width = 0, transparent = true) {
        super(null, color, width);
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
        this.id = id;
        this.darkAccent = COLOR.GREY;
        this.lightAccent = 'rgba(200, 200, 200, 255)';
        this.transparent = transparent;
        this.selected = false;
        this.coordinates = this.CalculatePoints();
        this.base = {
            "x": this.x,
            "y": this.y,
            "h": this.h,
            "w": this.w
        }
    }

    IsInside(position) {
        return this.x <= position[0] && this.x + this.w >= position[0] && this.y <= position[1] && this.y + this.h >= position[1];
    }

    DrawObject(screen) {
        super.coordinates = this.CalculatePoints();
        if (!this.transparent) {
            if (this.selected) {
                this.accentColor = this.darkAccent;
            } else {
                this.accentColor = this.lightAccent;
            }
        }
        super.DrawObject(screen);
    }

    GetId() {
        return this.id;
    }

    CalculatePoints() {
        var points = [];
        var percentToCut = 0.2;

        var tl = [this.x, this.y];
        var tr = [this.x + this.w, this.y];
        var bl = [this.x, this.y + this.h];
        var br = [this.x + this.w, this.y + this.h];
        var edgeCut = Math.trunc(this.h * percentToCut);

        points.push([tl[0], tl[1] + edgeCut])
        points.push([tl[0] + edgeCut, tl[1]])
        points.push([tr[0] - edgeCut, tr[1]])
        points.push([tr[0], tr[1] + edgeCut])
        points.push(br)
        points.push(bl)

        return points
    }

    UpdatePoints() {
        this.coordinates = this.CalculatePoints();
    }

    ResetPositions() {
        this.y = this.base.y;
        this.h = this.base.h;
        this.UpdatePoints();
    }

}